### PR TITLE
Update optimizer for correct compression optimization

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
It turns out we are adding quantizer and sparsity masks parameters to the optimizer group with weights decay. This lead to wrong sparsity training process. I've replaced new implementation to old one but it should work the same and it's more explicit as for me.

### CODE REVIEW

**Comment by daniil-lyakhov** [comment]
Tested locally

File: torchreid/optim/optimizer.py [hunk 1]
```diff
@@ -131,22 +135,53 @@ def _build_optim(model,
...
+                # Skip non trainable params
+                binary_masks_count += 1
+                continue
+            elif hasattr(m, 'weight') or hasattr(m, 'bias'):
+                if hasattr(m, 'weight'):
+                    weight_no_decay.append(m.weight)
+                if hasattr(m, 'bias'):
+                    bias_no_decay.append(m.bias)
+            elif len(list(m.children())) == 0:
+                for name, param in m.named_parameters():
```
Comment (kprokofi) [inline torchreid/optim/optimizer.py]: This won't work with transformer head

File: torchreid/optim/optimizer.py [hunk 2]
```diff
@@ -131,22 +135,53 @@ def _build_optim(model,
...
+                if m.bias is not None:
+                    bias_no_decay.append(m.bias)
+            elif isinstance(m, SymmetricQuantizer):
+                compression_params.append(m.scale)
+                fq_not_trainable_params_count += 2
+            elif isinstance(m, AsymmetricQuantizer):
+                compression_params.append(m.input_low)
+                compression_params.append(m.input_range)
+                fq_not_trainable_params_count += 1
+            elif isinstance(m, RBSparsifyingWeight):
```
Comment (kprokofi) [inline torchreid/optim/optimizer.py]: can you use parameters names?

File: torchreid/optim/optimizer.py (hunk 1 above)
Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py]: We should somehow aggregate compression parameters separate from others. We should discuss how to do this properly

File: torchreid/optim/optimizer.py [hunk 3]
```diff
@@ -4,6 +4,10 @@
 
 import warnings
 
+from nncf.torch.quantization.layers import SymmetricQuantizer
```
Comment (kprokofi) [inline torchreid/optim/optimizer.py]: we are using nncf as optional requirements
Do we have a way how to wrap these imports?

File: torchreid/optim/optimizer.py (hunk 2 above)
Comment (kprokofi) [reply torchreid/optim/optimizer.py]: We have not accidentally switched to a new type of handling with parameters.
These changes are important for the working of transformer heads
So, could you use our heuristic of parameters handling and treat nncf parameters through named_parameters?

Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py]: Yes, but I think this is more strict way to check is module compression module. I'll use modules attributes instead

File: torchreid/optim/optimizer.py (hunk 3 above)
Comment (kprokofi) [reply torchreid/optim/optimizer.py]: for example, you may look in engine.py. We wrap some imports

Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py]: Please check

File: torchreid/optim/optimizer.py (hunk 2 above)
Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py]: Done

File: torchreid/optim/optimizer.py [hunk 4]
```diff
@@ -143,10 +152,14 @@ def _build_optim(model,
             else:
                 decay.append(param)
 
-        assert len(list(model.parameters())) == len(decay) + len(bias_no_decay) + len(weight_no_decay)
```
Comment (daniil-lyakhov) [inline torchreid/optim/optimizer.py:146]: How it worked in case some params are have required_graph == false?

File: torchreid/optim/optimizer.py [hunk 5]
```diff
@@ -7,6 +7,12 @@
 from .sam import SAM
 from .radam import RAdam
 
+NNCF_ENABLED = True
+try:
+    from nncf.torch.layer_utils import CompressionParameter
+except ImportError:
+    NNCF_ENABLED = False
```
Comment (kprokofi) [inline torchreid/optim/optimizer.py]: what if the nncf is installed but we do not want to train the model with nncf?

Comment (kprokofi) [reply torchreid/optim/optimizer.py]: Please do a function that returns necessary imported modules and variables
https://github.com/openvinotoolkit/deep-object-reid/blob/ote/torchreid/integration/nncf/compression.py

File: torchreid/optim/optimizer.py (hunk 4 above)
Comment (kprokofi) [reply torchreid/optim/optimizer.py:146]: required_grad?

Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py:146]: *requires_grad

Comment (kprokofi) [reply torchreid/optim/optimizer.py:146]: As usual, I guess. The optimizer will not update those parameters with grad==False. I guess it doesn't matter if we pass all parameters to the optimizer together or separately by chanks

Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py:146]: I mean if one of the parameters will be `requires_grad==False` this assert will be activated

File: torchreid/optim/optimizer.py (hunk 5 above)
Comment (daniil-lyakhov) [reply torchreid/optim/optimizer.py]: Done

**Review by kprokofi — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
